packages/server/src/fhir/repo.ts

```diff
@@ -853,37 +865,42 @@ export class Repository extends BaseRepository implements FhirRepository<PoolCli
    */
   async forAllResources<T extends Resource>(
     resourceType: T['resourceType'],
-    callback: (resource: T) => Promise<void>
+    callback: (conn: PoolClient, resources: T[]) => Promise<void>
   ): Promise<void> {
-    const batchSize = 1000;
+    const batchSize = 500;
     let hasMore = true;
-    let currentTimestamp: string | undefined = undefined;
+    let currentTimestamp = new Date(0).toISOString(); // Beginning of epoch time
+    const endTimestamp = new Date(Date.now() + 1000 * 60 * 5).toISOString(); // Five minutes in the future
 
     while (hasMore) {
       const searchRequest: SearchRequest<T> = {
         resourceType,
         count: batchSize,
         sortRules: [{ code: '_lastUpdated', descending: false }],
+        filters: [
+          {
+            code: '_lastUpdated',
+            operator: Operator.GREATER_THAN_OR_EQUALS,
+            value: currentTimestamp,
+          },
+          { code: '_lastUpdated', operator: Operator.LESS_THAN, value: endTimestamp },
+        ],
       };
 
-      if (currentTimestamp) {
-        searchRequest.filters = [
-          { code: '_lastUpdated', operator: Operator.GREATER_THAN_OR_EQUALS, value: currentTimestamp },
-        ];
-      }
-
-      const bundle = await this.search(searchRequest);
-      if (bundle.entry) {
-        for (const entry of bundle.entry) {
-          const resource = entry.resource as T;
-          await callback(resource);
-
-          const lastUpdated = resource.meta?.lastUpdated as string;
-          currentTimestamp = lastUpdated;
+      await this.withTransaction(async (conn) => {
```

Comment:
Is this transaction necessary?  Trying to imagine what it's needed for, versus letting the callback handle it.

Reply:
Hm, I see, you're using this to avoid per-resource transactions in the reindex step.

imho, that kinda breaks the abstraction of `forAllResources`

1. I'm not sure the transaction is fully necessary in `reindexResource`
2. If it is both necessary and if this is a major performance win, then i think we should probably modify the API contract of `forAllResources` (maybe a "page" callback and a "resource" callback?)